drivers/random/random_nrf5.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2017 Exati Tecnologia Ltda.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <random.h>
+#include <atomic.h>
+#include <nrf_rng.h>
+
+struct random_nrf5_dev_data {
+	atomic_t user_count;
+};
+
+#define DEV_DATA(dev) \
+	((struct random_nrf5_dev_data *)(dev)->driver_data)
+
+static inline s16_t random_nrf5_get_u8(void)
+{
+	u8_t value;
+	u16_t timeout = 0;
+
+	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+		timeout++;
+		if (timeout >= 1000000) {
+			return -ETIMEDOUT;
+		}
+	}
+
+	value = nrf_rng_random_value_get();
+	nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+
+	return value;
+}
+
+static int random_nrf5_get_entropy(struct device *device, u8_t *buf, u16_t len)
+{
+	int ret = 0;
+
+	/* Mark the peripheral as being used */
+	atomic_inc(&DEV_DATA(device)->user_count);
+
+	/* Start the RNG generator peripheral */
+	nrf_rng_task_trigger(NRF_RNG_TASK_START);
+
+	while (len) {
+		s16_t value = random_nrf5_get_u8();
+
+		if (value < 0) {
+			ret = value;
+			break;
+		}
+
+		*buf = (u8_t) value;
+		buf++;
+		len--;
+	}
+
+	/* Release our use of the peripheral */
+	atomic_dec(&DEV_DATA(device)->user_count);
+
+	/* Only stop the RNG generator peripheral if we're the last user */
+	if (atomic_get(&DEV_DATA(device)->user_count) == 0) {
+		nrf_rng_task_trigger(NRF_RNG_TASK_STOP);
+	}
```

Comment:
What about using the VALRDY_STOP shortcut to turn off the peripheral, and getting rid of this timeout checking? I mean something like this:

```
atomic_inc(&DEV_DATA(device)->user_count);

nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
nrf_rng_task_trigger(NRF_RNG_TASK_START);

while (len) {
[..]
}

if (atomic_dec(&DEV_DATA(device)->user_count) == 1) {
	nrf_rng_shorts_enable(NRF_RNG_SHORT_VALRDY_STOP_MASK);

	if (atomic_get(&DEV_DATA(device)->user_count) != 0) {
		nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
		nrf_rng_task_trigger(NRF_RNG_TASK_START);
	}
}
```
The RNG peripheral will then be stopped after generating another random byte, unless some new request appears in the meantime.

This second `if` should solve the race condition when some higher priority code turns on the RNG peripheral and starts waiting for the VALRDY event right between `atomic_dec` and `nrf_rng_shorts_enable`.

Well, concurrency just complicates simple things... ;-)

Reply:
Concurrency: it is what it is, for better or worse. At least we can say we work doing hard things... :smile: 

I like this solution. I'll implement and test it as soon as I have some time. Thanks a lot, Andrzej.